### PR TITLE
Update mimir-prometheus to 320f0c9c4a88

### PR DESCRIPTION
**Pull request by aknuds1**
#### What this PR does
Update mimir-prometheus to latest main, i.e. [320f0c9c4a88](https://github.com/grafana/mimir-prometheus/tree/320f0c9c4a88f9c30a350900809ead6c302614ee).

The main change in Prometheus is that the context argument is moved from `storage.Queryable.Querier` and instead to individual query methods (`Select`, `LabelValues`, `LabelNames`). This requires some adapting of Mimir code, especially that which relies on the context being a querier field (e.g. tenant federation).

Also, due to (certain) Prometheus query methods gaining a context argument, they now respect context cancellation properly, unlike before.

**TODO**:
- [x] Add changes to the changelog
- [x] Verify whether query cancellation is properly implemented
- [ ] Add tests to verify that queries respect context cancellation?
- [x] Check query traces - we do some things to workaround the previous state to ensure the context is propagated correctly (cc @charleskorn)
- [x] Make sure tenant federation works with GEM (cc @pstibrany)

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir/issues/5314.

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/querier.go
Comment (aknuds1) [reply pkg/querier/querier.go]: I believe you are right @krajorama. Fixed.

File: pkg/querier/blocks_store_queryable_test.go
Comment (krajorama) [inline pkg/querier/blocks_store_queryable_test.go:1585]: nit: do it like on line 1508 for consistency at the top of this t.Run()

Comment (krajorama) [inline pkg/querier/blocks_store_queryable_test.go:1585]: nit: do it like on line 1508 for consistency at the top of this t.Run()

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (krajorama) [inline pkg/querier/tenantfederation/merge_queryable_test.go:285]: nit: moving it just made the diff bigger, otherwise no effect

File: pkg/storage/tsdb/block/index.go
Comment (krajorama) [inline pkg/storage/tsdb/block/index.go:627]: little bit weird that these 3 don't have context, but I guess these are not on the hotpath/weren't modified in Prometheus either

**Review by krajorama — COMMENTED** [review]
I've gone through it once. LGTM. The one thing I'm wondering about is where we need to add some tests to verify that the new functionality can actually cancel the requests we want to be able to cancel. To make sure that we've passed the right context in the right places.

File: pkg/ingester/ingester.go
Comment (aknuds1) [inline pkg/ingester/ingester.go:2217]: @pracucci made me aware that chances are it's a bad idea to pass a request-tied context to TSDB creation, since we don't want to interrupt the latter if the request gets canceled. Thinking the whole thing through, I suggest we don't pass down context to `createTSDB` at all in this PR (the context argument to `createTSDB` was previously unused). If there are cases for which we _want_ to let context cancellation interrupt TSDB creation (during service startup), it's probably better to enable it in a separate PR?

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable_test.go:285]: Yep, you're right. I'll undo this change.

**Comment by aknuds1** [comment]
> The one thing I'm wondering about is where we need to add some tests to verify that the new functionality can actually cancel the requests we want to be able to cancel

Good point @krajorama, I guess we should discuss this.

File: pkg/storage/tsdb/block/index.go
Comment (aknuds1) [reply pkg/storage/tsdb/block/index.go:627]: Correct, they weren't modified in Prometheus. Maybe we should review at a later point if there's a need.

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable_test.go]: Done.

File: pkg/ingester/ingester.go
Comment (aknuds1) [reply pkg/ingester/ingester.go:2217]: @colega my thinking too.

File: pkg/querier/querier.go
Comment (colega) [inline pkg/querier/querier.go:244]: Nit, but idiomatic goland way is to omit the `get` prefix calling this just `queriers()`

File: pkg/storegateway/indexheader/header.go
Comment (colega) [inline pkg/storegateway/indexheader/header.go:49]: Do we need the context on this one? I can't see where we need it upstream.

Comment (aknuds1) [reply pkg/storegateway/indexheader/header.go:49]: It's used in [index.Decoder.LookupSymbol](https://github.com/grafana/mimir/blob/fce0b89/pkg/storegateway/bucket_index_reader.go#L62), but I see that we can circumvent it as the context isn't used in our code. I can make a revision to remove the context argument.

Comment (aknuds1) [reply pkg/storegateway/indexheader/header.go]: Removed the context in latest revision, thanks for pointing this out.

Comment (colega) [inline pkg/storegateway/indexheader/header.go]: Do we need the context on this one? I can't see where we need it.

Comment (aknuds1) [reply pkg/storegateway/indexheader/header.go]: Gone now.

**Comment by colega** [comment]
LGTM, great work!

**Review by colega — APPROVED** [review]

**Comment by aknuds1** [comment]
Thank you @colega!

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pstibrany) [reply pkg/querier/tenantfederation/merge_queryable.go:54]: Removing support for `MergeQuerierCallback` also breaks our enterprise product.

I'd suggest to revert changes in this package and only do the necessary changes related to context propagation and annotations.

File: tools/tsdb-symbols/main.go
Comment (pstibrany) [inline tools/tsdb-symbols/main.go:92]: ha, how convenient, context was already there, just ignored :)

**Review by pstibrany — COMMENTED** [review]
Great work. I'm afraid that changes in `tenantfederation` package will break our enterprise product. I'd suggest to revert them, and only do changes related to Prometheus update.

I'd also suggest to try to integrate the changes into our enterprise product to be sure we don't break anything.

File: pkg/ingester/ingester.go
Comment (aknuds1) [reply pkg/ingester/ingester.go:2215]: The code prior to the PR calls `db.Compact` without a context though, so passing `context.Background()` here only retains the current (pre-PR) behaviour.

If we want to make compaction cancellable, we can do it in a follow-up PR (since otherwise the PR does yet another thing)?

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:54]: Aha thanks @pstibrany, I forgot to test my PR with enterprise. I should do that.

File: tools/tsdb-symbols/main.go
Comment (aknuds1) [reply tools/tsdb-symbols/main.go:92]: Yep, I did a preparation PR first :)

File: pkg/ingester/ingester.go
Comment (pstibrany) [reply pkg/ingester/ingester.go:2215]: > The code prior to the PR calls `db.Compact` without a context though, so passing `context.Background()` here only retains the current (pre-PR) behaviour.

Oh, I see. I would remove the comment then (I don't think it's true) and perhaps use `context.TODO()` as before.

> If we want to make compaction cancellable, we can do it in a follow-up PR (since otherwise the PR does yet another thing)?

Yes, let's do that in separate PR. Thanks.

File: tools/tsdb-symbols/main.go
Comment (pstibrany) [reply tools/tsdb-symbols/main.go:92]: Ah, right! That was your previous PR, I totally forgot.

**Comment by charleskorn** [comment]
> * Check query traces - we do some things to workaround the previous state to ensure the context is propagated correctly (cc @charleskorn)

Just checked this locally and traces look good to me, and it looks like this change has fixed #5314 as well 🎉

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pstibrany) [inline pkg/querier/tenantfederation/merge_queryable.go:59]: I'd suggest to add short doc string to the struct and fields.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pstibrany) [inline pkg/querier/tenantfederation/merge_queryable.go:96]: ... `by a MergeQuerierCallbacks.Querier callback.`

**Review by pstibrany — APPROVED** [review]
Thank you, `tenantfederation` changes make sense to me.

**Review by dimitarvdimitrov — COMMENTED** [review]
I managed to look over the changes in the ingester, tenantfederation, and frontend/* packages. Unfortunately my battery will die soon, so I will need to get back to the rest later.

There were quite a few stylistic changes that make this already large  change set even larger. For me at least this makes the review more error-prone. Perhaps doing them in a follow-up PR would have been easier to review. I don't think you should undo them at this point, but wanted to point this out

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:30]: @dimitarvdimitrov changing to `bypassWithSingleID` makes sense, since the number of queriers had to be abstracted away (only GEM will need multiple queriers). The doc strings need fixing up. I was waiting for Peter's input first :)

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: It's necessary to abstract since the enterprise code needs to do it differently (hence the callbacks).

**Comment by dimitarvdimitrov** [comment]
looks like others are happy with the change; i couldn't find any bugs in the code i've reviewed so far, so if you're confident, you can merge this without waiting for my full review (which might only come next week or on Sunday)

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:130]: Actually I am unsure. I was going to confer with @charleskorn, since I think it could be related to something he tested this PR for.

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable_test.go:364]: Do the federation IDs in our enterprise product, correspond to tenant IDs? I'm actually unsure about that part. That's why I generalized to "federation ID".

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable_test.go:167]: I will check.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:401]: Will check.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:334]: Will check, but will that not increase the changes in this PR?

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:343]: I will check, but I wonder if it will increase the amount of changes in the PR?

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:269]: Will check.

**Comment by aknuds1** [comment]
> There were quite a few stylistic changes that make this already large change set even larger. For me at least this makes the review more error-prone

@dimitarvdimitrov are you referring to the `tenantfederation` package? Changes were necessary there to support our enterprise product, which customizes federation logic.

File: pkg/mimir/modules.go
Comment (aknuds1) [reply pkg/mimir/modules.go:474]: You're right thanks, it was left behind from a previous change. Undid this change.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (charleskorn) [reply pkg/querier/tenantfederation/merge_queryable.go:130]: I didn't check queries with tenant federation specifically. However, I'd expect that this comment is no longer valid: we no longer need to do anything special to make sure spans are correctly linked to their parents because we can pass the context from the parent to the child.

Comment (charleskorn) [reply pkg/querier/tenantfederation/merge_queryable.go:130]: I've just tried a query with tenant federation and the traces look correct to me.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: This is fixed.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:269]: I'll do this simplification in a different PR, to reduce risk.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:343]: I'll do this in a separate PR, to reduce risk.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go:334]: Doing also this in a different PR, it stems from before this PR.

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable_test.go]: Done, thanks.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: Thank you, I've removed the TODO comment!

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: Done.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: Done.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: Done.

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable_test.go]: Fixed.

File: pkg/ingester/ingester.go
Comment (pracucci) [reply pkg/ingester/ingester.go:2217]: Agree with Oleg.

**Review by pracucci — APPROVED** [review]
Good job Arve! I can feel your pain while dealing with this PR 🤯 The changes to the query federation are quite involved, but I can't see anything wrong just looking at the diff.

I left few comments I would be glad if you could take a look before merging.

File: pkg/ingester/user_tsdb.go
Comment (aknuds1) [reply pkg/ingester/user_tsdb.go]: Done.

File: pkg/ingester/label_names_and_values.go
Comment (aknuds1) [reply pkg/ingester/label_names_and_values.go:34]: Done.

**Pull request merged**
